File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/tier/TierManagerTest.kt
package org.wyrdsekai.app.engine.tier

import kotlinx.coroutines.launch
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.yield
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class TierManagerTest {

    private fun probe(
        memoryMb: Long = 2000,
        battery: Int = 80,
        charging: Boolean = false,
        thermal: ThermalState = ThermalState.NOMINAL,
        wifi: Boolean = true,
    ) = object : ResourceProbe {
        override fun snapshot() = ResourceSnapshot(
            availableMemoryMb = memoryMb,
            totalMemoryMb = 4000,
            batteryPercent = battery,
            isCharging = charging,
            thermalState = thermal,
            hasWifi = wifi,
        )
    }

    @Test
    fun initializeSetsInitialTier() = runTest {
        val manager = TierManager(probe(memoryMb = 1500, wifi = false), scope = this)
        manager.initialize()
        assertEquals(Tier.T1, manager.currentTier.value)
    }

    @Test
    fun initializeWithLowResourcesSetsT0() = runTest {
        val manager = TierManager(probe(memoryMb = 500), scope = this)
        manager.initialize()
        assertEquals(Tier.T0, manager.currentTier.value)
    }

    @Test
    fun initializeWithHighResourcesSetsT3() = runTest {
        val manager = TierManager(probe(memoryMb = 4000, charging = true, wifi = true), scope = this)
        manager.initialize()
        assertEquals(Tier.T3, manager.currentTier.value)
    }

    @Test
    fun forceTierOverridesProbe() = runTest {
        val manager = TierManager(probe(memoryMb = 500), scope = this)
        manager.initialize()
        assertEquals(Tier.T0, manager.currentTier.value)

        manager.forceTier(Tier.T2)
        assertEquals(Tier.T2, manager.currentTier.value)
    }

    @Test
    fun tierTransitionsEmitted() = runTest {
        val manager = TierManager(probe(memoryMb = 1500, wifi = false), scope = this)
        val transitions = mutableListOf<TierTransition>()
        val job = launch { manager.transitions.collect { transitions.add(it) } }
        yield() // Let collector start

        manager.initialize()
        yield() // Let emission propagate
        // Initial transition T0 → T1
        assertEquals(1, transitions.size)
        assertEquals(Tier.T0, transitions[0].from)
        assertEquals(Tier.T1, transitions[0].to)
        assertEquals("initial", transitions[0].reason)

        manager.forceTier(Tier.T3)
        yield()
        assertEquals(2, transitions.size)
        assertTrue(transitions[1].isPromotion)

        manager.forceTier(Tier.T0)
        yield()
        assertEquals(3, transitions.size)
        assertTrue(transitions[2].isDemotion)

        job.cancel()
    }

    @Test
    fun configUpdatesWithTier() = runTest {
        val manager = TierManager(probe(), scope = this)
        manager.initialize()

        manager.forceTier(Tier.T0)
        assertEquals(InferenceMode.REMOTE, manager.config.value.inferenceMode)
        assertEquals(0, manager.config.value.maxRooms)

        manager.forceTier(Tier.T2)
        assertEquals(InferenceMode.LOCAL, manager.config.value.inferenceMode)
        assertEquals(4, manager.config.value.maxRooms)
        assertTrue(manager.config.value.betweenEnabled)
    }

    @Test
    fun monitoringDetectsChanges() = runTest {
        var currentMemory = 2000L
        val dynamicProbe = object : ResourceProbe {
            override fun snapshot() = ResourceSnapshot(
                availableMemoryMb = currentMemory,
                totalMemoryMb = 4000,
                batteryPercent = 80,
                isCharging = false,
                thermalState = ThermalState.NOMINAL,
                hasWifi = true,
            )
        }

        val manager = TierManager(dynamicProbe, probeIntervalMs = 1000, scope = this)
        val transitions = mutableListOf<TierTransition>()
        val job = launch { manager.transitions.collect { transitions.add(it) } }
        yield()

        manager.initialize()
        manager.startMonitoring()

        // Simulate memory drop
        currentMemory = 500
        advanceTimeBy(1500)
        yield()
        assertEquals(2, transitions.size)
        assertEquals(Tier.T0, transitions[1].to)

        // Simulate memory recovery
        currentMemory = 3000
        advanceTimeBy(1500)
        yield()
        assertEquals(3, transitions.size)
        assertEquals(Tier.T2, transitions[2].to)

        manager.stopMonitoring()
        job.cancel()
    }

    @Test
    fun noTransitionEmittedWhenTierUnchanged() = runTest {
        val manager = TierManager(probe(memoryMb = 1500, wifi = false), probeIntervalMs = 1000, scope = this)
        val transitions = mutableListOf<TierTransition>()
        val job = launch { manager.transitions.collect { transitions.add(it) } }
        yield()

        manager.initialize()
        manager.startMonitoring()

        // Advance several intervals — probe returns same resources, no new transitions
        advanceTimeBy(5000)
        yield()
        assertEquals(1, transitions.size) // Only the initial transition

        manager.stopMonitoring()
        job.cancel()
    }
}

File: clients/rn/__tests__/engine/tier-manager.test.ts
import type { ResourceProbe, ResourceSnapshot } from '../../src/engine/tier/TierConfig';
import { TierManager, type TierTransition, isPromotion, isDemotion } from '../../src/engine/tier/TierManager';

function makeProbe(overrides: Partial<ResourceSnapshot> = {}): ResourceProbe {
  return {
    snapshot: () => ({
      availableMemoryMb: 2000,
      totalMemoryMb: 4000,
      batteryPercent: 80,
      isCharging: false,
      thermalState: 'NOMINAL' as const,
      hasWifi: true,
      ...overrides,
    }),
  };
}

describe('TierManager', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('initialize sets initial tier', () => {
    const manager = new TierManager(makeProbe({ availableMemoryMb: 1500, hasWifi: false }));
    manager.initialize();
    expect(manager.currentTier).toBe('T1');
  });

  it('initialize with low resources sets T0', () => {
    const manager = new TierManager(makeProbe({ availableMemoryMb: 500 }));
    manager.initialize();
    expect(manager.currentTier).toBe('T0');
  });

  it('initialize with high resources sets T3', () => {
    const manager = new TierManager(makeProbe({
      availableMemoryMb: 4000, isCharging: true, hasWifi: true,
    }));
    manager.initialize();
    expect(manager.currentTier).toBe('T3');
  });

  it('forceTier overrides probe', () => {
    const manager = new TierManager(makeProbe({ availableMemoryMb: 500 }));
    manager.initialize();
    expect(manager.currentTier).toBe('T0');

    manager.forceTier('T2');
    expect(manager.currentTier).toBe('T2');
  });

  it('tier transitions emitted', () => {
    const manager = new TierManager(makeProbe({ availableMemoryMb: 1500, hasWifi: false }));
    const transitions: TierTransition[] = [];
    manager.onTransition(t => transitions.push(t));

    manager.initialize();
    // Initial transition T0 → T1
    expect(transitions.length).toBe(1);
    expect(transitions[0].from).toBe('T0');
    expect(transitions[0].to).toBe('T1');
    expect(transitions[0].reason).toBe('initial');

    manager.forceTier('T3');
    expect(transitions.length).toBe(2);
    expect(isPromotion(transitions[1])).toBe(true);

    manager.forceTier('T0');
    expect(transitions.length).toBe(3);
    expect(isDemotion(transitions[2])).toBe(true);
  });

  it('config updates with tier', () => {
    const manager = new TierManager(makeProbe());
    manager.initialize();

    manager.forceTier('T0');
    expect(manager.config.inferenceMode).toBe('REMOTE');
    expect(manager.config.maxRooms).toBe(0);

    manager.forceTier('T2');
    expect(manager.config.inferenceMode).toBe('LOCAL');
    expect(manager.config.maxRooms).toBe(4);
    expect(manager.config.betweenEnabled).toBe(true);
  });

  it('monitoring detects changes', () => {
    let currentMemory = 2000;
    const dynamicProbe: ResourceProbe = {
      snapshot: () => ({
        availableMemoryMb: currentMemory,
        totalMemoryMb: 4000,
        batteryPercent: 80,
        isCharging: false,
        thermalState: 'NOMINAL' as const,
        hasWifi: true,
      }),
    };

    const manager = new TierManager(dynamicProbe, 1000);
    const transitions: TierTransition[] = [];
    manager.onTransition(t => transitions.push(t));

    manager.initialize();
    manager.startMonitoring();

    // Simulate memory drop
    currentMemory = 500;
    jest.advanceTimersByTime(1500);
    expect(transitions.length).toBe(2);
    expect(transitions[1].to).toBe('T0');

    // Simulate memory recovery
    currentMemory = 3000;
    jest.advanceTimersByTime(1500);
    expect(transitions.length).toBe(3);
    expect(transitions[2].to).toBe('T2');

    manager.stopMonitoring();
  });

  it('no transition emitted when tier unchanged', () => {
    const manager = new TierManager(
      makeProbe({ availableMemoryMb: 1500, hasWifi: false }),
      1000,
    );
    const transitions: TierTransition[] = [];
    manager.onTransition(t => transitions.push(t));

    manager.initialize();
    manager.startMonitoring();

    // Advance several intervals — probe returns same resources
    jest.advanceTimersByTime(5000);
    expect(transitions.length).toBe(1); // Only the initial transition

    manager.stopMonitoring();
  });

  it('unsubscribe stops receiving transitions', () => {
    const manager = new TierManager(makeProbe());
    const transitions: TierTransition[] = [];
    const unsub = manager.onTransition(t => transitions.push(t));

    manager.initialize();
    expect(transitions.length).toBe(1);

    unsub();
    manager.forceTier('T3');
    expect(transitions.length).toBe(1); // No new transition received
  });
});
